src/entities/database.js

```diff
@@ -268,26 +305,170 @@ class DB implements Database {
       tx_ordinal: tx.txOrdinal,
       time: txUTCTime,
       last_update: txUTCTime,
+      ...(!_.isEmpty(inputUtxos)
+        ? {
+          inputs: JSON.stringify(inputUtxos),
+          inputs_address: inputAddresses,
+          inputs_amount: inputAmmounts,
+        }
+        : {}),
     }
     const now = new Date().toUTCString()
-    const query = Q.TX_INSERT.setFields(txDbFields)
-      .onConflict('hash', {
+
+    const onConflictArgs = []
+    if (upsert) {
+      onConflictArgs.push('hash', {
         block_num: blockNum,
         block_hash: blockHash,
         time: txUTCTime,
         tx_state: txStatus,
         last_update: now,
         tx_ordinal: tx.txOrdinal,
       })
+    }
+
+    const sql = Q.TX_INSERT.setFields(txDbFields)
+      .onConflict(...onConflictArgs)
       .toString()
-    this.#logger.debug('Insert TX:', query, inputAddresses, inputAmmounts)
-    await conn.query(query)
+    this.#logger.debug('Insert TX:', sql, inputAddresses, inputAmmounts)
+    await conn.query(sql)
     await this.storeTxAddresses(
       id,
       [...new Set([...inputAddresses, ...outputAddresses])],
     )
   }
 
+  async isTxExists(txId: string): boolean {
+    const sql = Q.sql.select().from('txs')
+      .field('1')
+      .where('hash = ?', txId)
+      .limit(1)
+      .toString()
+    const dbRes = await this.getConn().query(sql)
+    return dbRes.rows.length === 1
+  }
+
+  async isTxValid(txHash: string): Promise<boolean> {
+    const sql = Q.sql.select().from('txs')
+      .where('hash = ?', txHash)
+      .where('tx_state = ?', TX_STATUS.TX_SUCCESS_STATUS)
+      .toString()
+    const dbRes = await this.getConn().query(sql)
+    const [tx] = dbRes.rows
+    return Boolean(tx) && this.utxosForInputsExists(tx.inputs)
+  }
+
+  async queryPendingSet() {
+    const query = Q.sql.select().from(SNAPSHOTS_TABLE)
+      .field('tx_hash')
+      .where('block_height = ?', Q.sql.select().from(SNAPSHOTS_TABLE)
+        .field('MAX(block_height)'))
+      .where('status = ?', TX_STATUS.TX_PENDING_STATUS)
+      .union(Q.sql.select().from('txs')
+        .field('hash')
+        .where('tx_state = ?', TX_STATUS.TX_PENDING_STATUS)
+        .where('NOT EXISTS ?', Q.sql.select().from(SNAPSHOTS_TABLE)
+          .field('1')
+          .where('tx_hash = hash')))
+      .toString()
+    const dbRes = await this.getConn().query(query)
+    this.#logger.debug('queryPendingSet:', query, dbRes)
+    return _.map(dbRes.rows, 'tx_hash')
+  }
+
+  async groupPendingTxs(txs: Array<string>): Promise<[Array<string>, Array<string>]> {
+    const validTxs = []
+    const invalidTxs = []
+    for (const tx of txs) {
+      const isValid = await this.isTxValid(tx)
+      if (isValid) {
+        validTxs.push(tx)
+      } else {
+        this.#logger.info(`tx ${tx} inputs already spent`)
+        invalidTxs.push(tx)
+      }
+    }
+    return [validTxs, invalidTxs]
+  }
```

Comment:
This function is a bit misleading, because `isTxValid` for example will return `false` even if tx exists, but already has status "Failed" or "Successful".

I think the logic should be changed a bit as something like:
1. Take argument as an array of tx hashes
2. Query inputs and statuses for all hashes (might be a batch)
3. Filter out any non-pending records
4. Validate other records and put into different groups

Reply:
Logically we only send tx-hashes here that we already know are pending at the moment, but just that logic by itself that anything that's *NOT "Pending AND Valid"* is considered invalid is kinda confusing. I think `isTxValid` can be replaced with a function like `selectInputsForPendingTxsOnly` which would take a list of hashes and select only rows where status is "Pending", then this function could be changed to something like:
```
groupPendingTxs(txHashes) {
   const inputArrays: Array<Array<any>> = selectInputsForPendingTxsOnly(txHashes)
   for (const inputArray of inputArrays) {
      if (this. utxosForInputsExists(inputArray)) {
         // add to pending
      } else {
         // add to failed
      }
   }
}
```

as one of the options